_posts/2020-03-15-method-s-arguments-pt-1.markdown

```diff
@@ -0,0 +1,372 @@
+```
+
+Also, we don't need to have simple operations like that for the default value, we can assign a function that will be called when assigning the values:
+
+```ruby
+def two_times(x)
+  x * 2
+end
+
+def foo(arg1 = 1, arg2 = two_times(arg1))
+  puts "arg1 is: #{arg1.inspect}"
+  puts "arg2 is: #{arg2.inspect}"
+end
+
+foo
+# => arg1 is: 1
+#    arg2 is: 2
+
+foo(5)
+# => arg1 is: 5
+#    arg2 is: 10 # it's multiplying 5, not 1, it uses our value, not te default one
+
+foo(5, 6)
+# => arg1 is: 5
+#    arg2 is: 6 # we are actually overriding that multiplication
+```
+
+# Variable arguments
+
+This type of optional positional arguments don't have a default value. They exist only if assigned and you can access them using a special array of arguments:
+
+```ruby
+def foo(*args) # we use this special syntax with the * (splat operator) at the beginning of the paremeter
+  puts "args is: #{args.inspect}"
+end
+
+
+foo(1)
+# => args is: [1]
+
+foo(1,2)
+# => args is: [1, 2]
+```
+
+You can combine this with required and optional arguments too:
+
+```ruby
+def foo(arg1, arg2 = 'default value', *args)
+  puts "arg1 is: #{arg1.inspect}"
+  puts "arg2 is: #{arg2.inspect}"
+  puts "args is: #{args.inspect}"
+end
+
+foo(1, 2, 3, 4)
+# => arg1 is: 1
+#    arg2 is: 2
+#    args is: [3, 4]
+
+foo(1, 2, 3)
+# => arg1 is: 1
+#    arg2 is: 2
+#    args is: [3]
+
+foo(1, 2)
+# => arg1 is: 1
+#    arg2 is: 2
+#    args is: []
+
+foo(1)
+# => arg1 is: 1
+#    arg2 is: default value
+#    args is: []
+```
+
+Notice that "args" is always an array, you usually deconstruct the array or us it as is as an array or to call another method:
+
+```ruby
+def bar(arg1, *args)
+  puts "arg1 in bar is: #{arg1.inspect}"
+  puts "args in bar is: #{args.inspect}"
+end
+
+def foo(arg1, *args)
+  puts "arg1 in foo is: #{arg1.inspect}"
+  puts "args in foo is: #{args.inspect}"
+
+  # desconstruct the array
+  dec1, dec2, *rest = args
+  puts "dec1 is: #{dec1.inspect}"
+  puts "dec2 is: #{dec2.inspect}"
+  puts "rest is: #{rest.inspect}"
+
+  # we need the * again so ruby deconstructs the array in different
+  # if we don't do this, args would be an array assigned to arg1
+  bar(*args)
+end
+
+foo(1, 2, 3, 4)
+# => arg1 in foo is: 1
+#    arg2 in foo is: [2, 3, 4]
+#    dec1 is: 2
+#    dec2 is: 3
+#    rest is: [4]
+#    arg1 in bar is: 2
+#    args in bar is: [3, 4]
+```
+
+> This special parameter must always be used after the other positional arguments.
+
+# Conclusion
+
+All these options gives us some flexibility, but the order being that rigid also forces us to remember and respect the order and also remember what does each position is, which really limits us.
+
+In the next post of this series we'll talk about keyword arguments that allows us to set arguments in any order, and on the next one we'll talk about the special block argument to inject code, some extra features like array desconstruction parameters and partially applied arguments to generate other methods.
```

Comment:
```suggestion
In the next post of this series we'll talk about keyword arguments that allow us to set arguments in any order, and on the next one we'll talk about the special block argument to inject code and some extra features like array deconstruction parameters and partially applied arguments to generate other methods.
```

Reply:
hmmm I don't like having more than one 'and' with no comma, maybe something like:

"...special block argument to inject code, some extra features like array deconstruction parameters, and partially applied arguments..."

?